makeflow/src/makeflow_hook.c

```diff
@@ -33,47 +33,68 @@ struct list * makeflow_hooks = NULL;
-    char * name_on_execution = string_replace_percents(name_on_execution_pattern, id);
+	char *id = string_format("%d",task->taskid);
+	char * name_on_submission = string_replace_percents(name_on_submission_pattern, id);
+	char * name_on_execution = string_replace_percents(name_on_execution_pattern, id);
 
-    /* Output of dag_file is returned to use for final filename. */
-    struct dag_file *f = dag_file_lookup_or_create(d, name_on_submission);
+	/* Output of dag_file is returned to use for final filename. */
+	struct dag_file *f = dag_file_lookup_or_create(d, name_on_submission);
 
-    batch_task_add_input_file(task, name_on_submission, name_on_execution);
+	batch_task_add_input_file(task, name_on_submission, name_on_execution);
 
-    free(id);
-    free(name_on_submission);
-    free(name_on_execution);
+	free(id);
+	free(name_on_submission);
+	free(name_on_execution);
 
-    return f;
+	return f;
 }
 
 struct dag_file * makeflow_hook_add_output_file(struct dag *d, struct batch_task *task, const char * name_on_submission_pattern, const char * name_on_execution_pattern)
 {
-    char *id = string_format("%d",task->taskid);
-    char * name_on_submission = string_replace_percents(name_on_submission_pattern, id);
-    char * name_on_execution = string_replace_percents(name_on_execution_pattern, id);
+	char *id = string_format("%d",task->taskid);
+	char * name_on_submission = string_replace_percents(name_on_submission_pattern, id);
+	char * name_on_execution = string_replace_percents(name_on_execution_pattern, id);
 
-    /* Output of dag_file is returned to use for final filename. */
-    struct dag_file *f = dag_file_lookup_or_create(d, name_on_submission);
+	/* Output of dag_file is returned to use for final filename. */
+	struct dag_file *f = dag_file_lookup_or_create(d, name_on_submission);
 
-    batch_task_add_output_file(task, name_on_submission, name_on_execution);
+	batch_task_add_output_file(task, name_on_submission, name_on_execution);
 
-    free(id);
-    free(name_on_submission);
-    free(name_on_execution);
+	free(id);
+	free(name_on_submission);
+	free(name_on_execution);
 
-    return f;
+	return f;
 }
 
 
-void makeflow_hook_register(struct makeflow_hook *hook) {
+int makeflow_hook_register_hook(struct makeflow_hook *hook) {
 	assert(hook);
 	if (!makeflow_hooks) makeflow_hooks = list_create();
-	struct makeflow_hook *h = xxmalloc(sizeof(*h));
-	memcpy(h, hook, sizeof(*h));
-	list_push_head(makeflow_hooks, h);
+
+	/* Add hook by default, if it doesn't exists in list of hooks. */
+	int rc = MAKEFLOW_HOOK_SUCCESS;
+	struct makeflow_hook *h = NULL;
+
+	if(hook->register_hook){
+		rc = hook->register_hook(hook, makeflow_hooks);
+	} else {
+		list_first_item(makeflow_hooks);
+		while((h = list_next_item(makeflow_hooks))){
+			if(!strcmp(h->module_name, hook->module_name)){
+				rc = MAKEFLOW_HOOK_SKIP;
+				break;
+			}
+		}
+	}
+
+	if(rc == MAKEFLOW_HOOK_SUCCESS){
```

Comment:
Would we want to check for failure here? `MAKEFLOW_HOOK_SKIP` should fall through, but modules might want to return failure and abort Makeflow if there's a very serious incompatibility.

Reply:
This was addressed, git hub did not detect to difference as the change is several lines later.